src/nvim/eval.c

```diff
@@ -20477,6 +20477,21 @@ void set_selfdict(typval_T *rettv, dict_T *selfdict)
   }
 }
 
+// Turn a typeval into a string.  Similar to tv_get_string_buf() but uses
+// string() on Dict, List, etc.
+static const char *tv_stringify(typval_T *varp, char *buf)
+  FUNC_ATTR_NONNULL_ALL
+{
+  if (varp->v_type == VAR_LIST
+      || varp->v_type == VAR_DICT
+      || varp->v_type == VAR_FUNC
+      || varp->v_type == VAR_PARTIAL
+      || varp->v_type == VAR_FLOAT) {
+    return encode_tv2string(&varp[0], NULL);
```

Comment:
ah, good idea. but why `&varp[0]` instead of just `varp`?

Also we still need to free `varp` before return the result. 

    tv_clear_(varp);

Reply:
I copy-pasted from `f_string`. `varp` is cleared at the bottom of the while loop.

```c
    // ex_execute while loop
    ret = eval1_emsg(&arg, &rettv, !eap->skip);
    if (ret == FAIL) {
      break;
    }

    if (!eap->skip) {
      char buf[NUMBUFLEN];
      const char *const argstr = eap->cmdidx == CMD_execute
        ? tv_get_string_buf(&rettv, buf)
        : tv_stringify(&rettv, buf);
      const size_t len = strlen(argstr);
      ga_grow(&ga, len + 2);
      if (!GA_EMPTY(&ga)) {
        ((char_u *)(ga.ga_data))[ga.ga_len++] = ' ';
      }
      memcpy((char_u *)(ga.ga_data) + ga.ga_len, argstr, len + 1);
      ga.ga_len += len;
    }

    tv_clear(&rettv);
```

Reply:
asan found a memory leak.

Reply:
nvrmnd. There's no perf issues anyway.